File: packages/engine/src/cpus/6502/instructions/lsr.js
import { AddressingMode, CycleType } from '..'

/**
 * Logical Shift Right
 *
 * A,C,Z,N = A/2 or M,C,Z,N = M/2
 *
 * Affects Flags: N Z C
 *
 * @example
 * MODE          SYNTAX        HEX LEN TIM
 * Accumulator   LSR A         $4A  1   2
 * Zero Page     LSR $44       $46  2   5
 * Zero Page,X   LSR $44,X     $56  2   6
 * Absolute      LSR $4400     $4E  3   6
 * Absolute,X    LSR $4400,X   $5E  3   7
 */
export function* lsr(mode, addr) {
  if (mode === AddressingMode.ACC) {
    this.carry = (this.acc & 0x1) > 0
    this.acc >>>= 1
    this.zero = this.acc === 0
    this.negative = (this.acc & 0x80) > 0
    yield CycleType.GET
  } else {
    if (Array.isArray(addr) && !addr[1]) {
      this.read(addr[0])
      yield CycleType.GET
    }

    addr = Array.isArray(addr) ? addr[0] : addr
    let b = this.read(addr)
    yield CycleType.GET

    this.write(addr, b)
    yield CycleType.PUT

    this.carry = (b & 0x1) > 0
    b >>>= 1
    this.write(addr, b)
    this.zero = b === 0
    this.negative = (b & 0x80) > 0
    yield CycleType.PUT
  }
}

File: packages/engine/src/cpus/6502/instructions/lsr.zig
const std = @import("std");
const CPU = @import("../cpu.zig").CPU;
const Addressing = @import("../opcode.zig").Addressing;

// Logical Shift Right
//
// A,C,Z,N = A/2 or M,C,Z,N = M/2
//
// Affects Flags: N Z C
//
// MODE          SYNTAX        HEX LEN TIM
// Accumulator   LSR A         $4A  1   2
// Zero Page     LSR $44       $46  2   5
// Zero Page,X   LSR $44,X     $56  2   6
// Absolute      LSR $4400     $4E  3   6
// Absolute,X    LSR $4400,X   $5E  3   7

pub fn lsr(self: *CPU, addressing: Addressing) void {
    const S = struct {
        var address: u16 = 0;
        var value: u8 = 0;
    };

    switch (addressing) {
        .acc => |_| {
            self.c = (self.acc & 0x1) > 0;
            self.acc >>= 1;
            self.z = self.acc == 0;
            self.n = (self.acc & 0x80) > 0;
            return;
        },
        .zpg => |v| {
            S.address = v;
            if (self.cycle_counter == 3) {
                S.value = self.read(v);
                self.next_cycle = .write;
                return;
            } else if (self.cycle_counter == 4) {
                self.write(v, S.value);
                self.next_cycle = .write;
                return;
            }
        },
        .zpx, .zpy => |v| {
            S.address = v;
            if (self.cycle_counter == 4) {
                S.value = self.read(v);
                self.next_cycle = .write;
                return;
            } else if (self.cycle_counter == 5) {
                self.write(v, S.value);
                self.next_cycle = .write;
                return;
            }
        },
        .abs => |v| {
            S.address = v;
            if (self.cycle_counter == 4) {
                S.value = self.read(v);
                self.next_cycle = .write;
                return;
            } else if (self.cycle_counter == 5) {
                self.write(v, S.value);
                self.next_cycle = .write;
                return;
            }
        },
        .abx, .aby => |v| {
            S.address = v.@"0";
            if (self.cycle_counter == 4) {
                _ = self.read(v.@"0");
                self.next_cycle = .read;
                return;
            } else if (self.cycle_counter == 5) {
                S.value = self.read(v.@"0");
                self.next_cycle = .write;
                return;
            } else if (self.cycle_counter == 6) {
                self.write(v.@"0", S.value);
                self.next_cycle = .write;
                return;
            }
        },
        else => {},
    }

    var b = S.value;
    self.c = (b & 0x1) > 0;
    b >>= 1;
    self.write(S.address, b);
    self.z = b == 0;
    self.n = (b & 0x80) > 0;
}

test {
    _ = @import("lsr.test.zig");
}
